daffodil-lib/src/main/scala/org/apache/daffodil/util/Validators.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.util
+
+import java.util.ServiceLoader
+
+import org.apache.daffodil.api.Validator
+
+object Validators {
+  import scala.collection.JavaConverters._
+
+  private val impls = new ThreadLocal[Map[String, Validator]] {
```

Comment:
This may have been over engineered during the first pass, likely as a reaction to the schema cache in the original Validator.

Can probably simplify here with a hash map and say stateless thread safe validators.  Thoughts?

Reply:
When DataProcessor.parse is called and validation is on, those calls are allowed to happen on different threads, so yes, I think validation can happen simultaneously on mulitple threads.

Reply:
Sounds good, will leave it as is; added a comment

Reply:
I would expect this to not need a ThreadLocal. I would expect that if a Validator is not thread-safe then the internals of that implementation should have to handle the ThreadLocals. This way, we only incur the overhead of ThreadLocals if a particular implementation actually needs it.

Reply:
I think the thread safety here guards Daffodil as well, where you could have a race at init time of the Validators state.

After sleeping on it I like the thread local here, but if it doesnt make sense Ill pull it out, you fellas know the daffodil flow better than I.

Reply:
Thinking about alternatives, will get something pushed for discussion later today.

Reply:
Pushed a commits that attempts to resolve this discussion using the validation compiler by making loading more explicit and and reducing the scope of the caching, letting it up to the client how to share things.